bundle/config/mutator/process_target_mode.go

```diff
@@ -138,19 +139,13 @@ func validateProductionMode(ctx context.Context, b *bundle.Bundle, isPrincipalUs
 	return nil
 }
 
-// Determines whether a service principal identity is used to run the CLI.
-func isServicePrincipalUsed(ctx context.Context, b *bundle.Bundle) (bool, error) {
-	ws := b.WorkspaceClient()
-
-	// Check if a principal with the current user's ID exists.
-	// We need to use the ListAll method since Get is only usable by admins.
-	matches, err := ws.ServicePrincipals.ListAll(ctx, iam.ListServicePrincipalsRequest{
-		Filter: "id eq " + b.Config.Workspace.CurrentUser.Id,
-	})
-	if err != nil {
-		return false, err
-	}
-	return len(matches) > 0, nil
+// Determines whether a given user id is a service principal.
+// This function uses a heuristic: if no user exists with this id, we assume
+// it's a service principal. Unfortunately, the standard service principal API is too
+// slow for our purposes.
+func IsServicePrincipal(ctx context.Context, ws *databricks.WorkspaceClient, userId string) bool {
```

Comment:
For both of these places we call it on-demand though. The call takes time an it may fail. And for `databricks bundle init`, we don't have a bundle with CurrentUser at our disposal.

Reply:
I see. My main concern is that we don't cache result of `is_service_principal` template helper function and if it's used multiple times in template it will result in multiple API calls. Shall we then just cache the result of the call in `is_service_principal ` similar to how we do with `user_name` function?

Reply:
Ack. Added caching!